presto-router/src/main/java/com/facebook/presto/router/cluster/ClusterManager.java

```diff
@@ -13,81 +13,205 @@
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.stream.Collectors;
 
+import static com.facebook.airlift.concurrent.Threads.threadsNamed;
 import static com.facebook.presto.router.RouterUtil.parseRouterConfig;
-import static com.facebook.presto.router.scheduler.SchedulerType.WEIGHTED_RANDOM_CHOICE;
 import static com.facebook.presto.spi.StandardErrorCode.CONFIGURATION_INVALID;
 import static com.google.common.base.Preconditions.checkArgument;
 import static com.google.common.collect.ImmutableList.toImmutableList;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
 import static java.util.stream.Collectors.toMap;
 
 public class ClusterManager
 {
-    private final Map<String, GroupSpec> groups;
-    private final List<SelectorRuleSpec> groupSelectors;
-    private final SchedulerType schedulerType;
-    private final Scheduler scheduler;
-    private final HashMap<String, HashMap<URI, Integer>> serverWeights = new HashMap<>();
+    private AtomicReference<ClusterManagerConfig> currentConfig;
+
+    private HashMap<String, HashMap<URI, Integer>> serverWeights = new HashMap<>();
+    private HashMap<URI, URI> discoveryURIs = new HashMap<>();
+    public final RouterConfig routerConfig;
+    private final RemoteInfoFactory remoteInfoFactory;
+    private final Logger log = Logger.get(ClusterManager.class);
+
+    // Cluster status
+    private static Duration pollingInterval;
+    private final ConcurrentHashMap<URI, RemoteClusterInfo> remoteClusterInfos = new ConcurrentHashMap<>();
+    private final ConcurrentHashMap<URI, RemoteQueryInfo> remoteQueryInfos = new ConcurrentHashMap<>();
+
+    private AtomicBoolean isWatchServiceStarted = new AtomicBoolean();
+    public OnConfigChangeDetection onConfigChangeDetection;
 
     @Inject
-    public ClusterManager(RouterConfig config)
+    public ClusterManager(RouterConfig config, RemoteInfoFactory remoteInfoFactory, RemoteStateConfig remoteStateConfig)
     {
+        this.routerConfig = config;
         RouterSpec routerSpec = parseRouterConfig(config)
                 .orElseThrow(() -> new PrestoException(CONFIGURATION_INVALID, "Failed to load router config"));
+        Map<String, GroupSpec> groups = ImmutableMap.copyOf(routerSpec.getGroups().stream().collect(toMap(GroupSpec::getName, group -> group)));
+        List<SelectorRuleSpec> groupSelectors = ImmutableList.copyOf(routerSpec.getSelectors());
+        Scheduler scheduler = new SchedulerFactory(routerSpec.getSchedulerType()).create();
+        this.currentConfig = new AtomicReference<>(new ClusterManagerConfig(groups, groupSelectors, scheduler));
+        this.remoteInfoFactory = requireNonNull(remoteInfoFactory, "remoteInfoFactory is null");
+        this.initializeServerWeights();
```

Comment:
Instead of duplicating this call, just call `onConfigChangeDetection.apply()` after defining it ?